src/storage/src/render/upsert.rs

```diff
@@ -434,8 +485,34 @@ where
         }
     });
 
-    output.as_collection().map(|result| match result {
-        Ok(ok) => Ok(ok),
-        Err(err) => Err(DataflowError::from(EnvelopeError::Upsert(err))),
-    })
+    (
+        output.as_collection().map(|result| match result {
+            Ok(ok) => Ok(ok),
+            Err(err) => Err(DataflowError::from(EnvelopeError::Upsert(err))),
+        }),
+        health_stream,
+    )
+}
+
+async fn process_upsert_state_error<G: Scope>(
+    msg: &str,
+    e: anyhow::Error,
+    worker_id: usize,
+    health_output: &mut AsyncOutputHandle<
+        <G as ScopeParent>::Timestamp,
+        Vec<(usize, usize, HealthStatusUpdate)>,
+        TeeCore<<G as ScopeParent>::Timestamp, Vec<(usize, usize, HealthStatusUpdate)>>,
+    >,
+    health_cap: &Capability<<G as ScopeParent>::Timestamp>,
+) {
+    let update = HealthStatusUpdate {
+        update: HealthStatus::StalledWithError {
+            error: format!("{}: {}", msg, e.display_with_causes()),
```

Comment:
```suggestion
    context: &str,
    e: anyhow::Error,
    worker_id: usize,
    health_output: &mut AsyncOutputHandle<
        <G as ScopeParent>::Timestamp,
        Vec<(usize, usize, HealthStatusUpdate)>,
        TeeCore<<G as ScopeParent>::Timestamp, Vec<(usize, usize, HealthStatusUpdate)>>,
    >,
    health_cap: &Capability<<G as ScopeParent>::Timestamp>,
) {
    let update = HealthStatusUpdate {
        update: HealthStatus::StalledWithError {
            error: e.context(context).to_string_with_causes(),
```

If you import https://docs.rs/anyhow/latest/anyhow/trait.Context.html, this is much cleaner, in my opinion!